### DISCUSSION TITLE
How do I trigger `re-frame/reg-fx` with this library?

### DISCUSSION THREAD

**Discussion by Invertisment** [day8/re-frame-async-flow-fx#36] [Q&A]
I'm new to re-frame and I want to control a pipeline of promises.

I think this library allows to do it but it requires to register EVent handlers for each action and I want to use EFfects.
I think that using EFfects is better as they don't mutate the application state and instead trigger EFfects which I'd like to listen to.

The tutorial of re-frame tells that EFfects are what should be used mutations (outside world, data fetching, etc.) and EVents (the ones who are triggered via `dispatch`) should be used to process user's callbacks and change application's state.
This is the doc: https://day8.github.io/re-frame/Effects/.

I'd like to know what is the best way to use this library if I want to fire several consecutive EFfects which do some fetching. The original re-frame tutorial doesn't show any example which deals with execution of two consecutive requests, for instance.

I think I simply need an example because I can't think of a better way than to wrap every step of my logic into a EVent handler that requests EFfects to happen. Is this the way to do it?
(By "wrap" I mean to have an EVent that simply returns cofx that triggers one or multiple EFfects)

**Comment by mike-thompson-day8** [comment]
Issues are for bugs and feature requests. 

Best to get tutorial help via slack (click on this badge) or Github's discussions feature :
[![Get help on Slack](http://img.shields.io/badge/slack-clojurians%20%23re--frame-97C93C?labelColor=283C67&logo=slack&style=for-the-badge)](https://clojurians.slack.com/channels/re-frame)

In the meantime, you might be interested to have a look at the `re-frame-http-fx` library which has to initiate async actions and then handle success or failure.

**Reply by mike-thompson-day8** [reply]
Sorry, when I originally answered this question, I thought you had asked it in the `re-frame` repo, which is why I simply pushed you towards `re-frame-http-fx`.

But now I can see you asked it here in `re-frame-async-flow-fx`.  So, I need to refine my answer ...

This library concerns itself only with coordinating the firing of events.  "When this event A happens, then fire event B". It is the handlers for those events which do the heavy lifting of returning effects, which, in turn, get handled by effect handlers.

So, your promises will have to dispatch events when they resolve (see `re-frame-http-fx`), and this library could then notices the events being dispatched and coordinate the further dispatch of other events.

**Reply by Invertisment** [reply]
The HTTP using promises was only an example. I don't want to know what kind of requests are done via the library. In reality I'm dealing with a library that returns promises from its method calls. And they don't even say that they use HTTP or anything else.

I want to contain something like this example in JS (I'll need to do something more complicated but this describes the problem):
```
() => {
  const permissionInfo = await jslib.requestPermissionAndInfo("arg for 1");
  const result = await jslib.fetchSomething(permissionInfo.dataField, "arg for 2");
  console.log(result) // This is what should end up in :db
}
```
I think that what re-frame asks for is to isolate both promise creation calls into their own EFfects. I did that but now there's this amount of complexity that I have to deal with. I can hack away by bundling these calls into one EFfect but I don't know what's important now. Re-frame docs tell that EFfects should be as simple as possible.

If I understand correctly then every EFfect should have an EVent handler. And if I want to have two futures then it becomes 5 chunks of code (2 EFfects, 2 dummy EVents, 1 pipeline described via this library).
I wanted to get rid of the EVents as they only trigger one thing each. I know that I can wrap that into a factory method but I wanted to not have them in the first place.
